airflow/gcp/hooks/pubsub.py

```diff
@@ -0,0 +1,300 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+This module contains a Google Pub/Sub Hook.
+"""
+
+from uuid import uuid4
+
+from googleapiclient.discovery import build
+from googleapiclient.errors import HttpError
+
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+
+
+def _format_subscription(project, subscription):
+    return 'projects/{}/subscriptions/{}'.format(project, subscription)
+
+
+def _format_topic(project, topic):
+    return 'projects/{}/topics/{}'.format(project, topic)
+
+
+class PubSubException(Exception):
+    """
+    Alias for Exception.
+    """
+
+
+class PubSubHook(GoogleCloudBaseHook):
+    """
+    Hook for accessing Google Pub/Sub.
+
+    The GCP project against which actions are applied is determined by
+    the project embedded in the Connection referenced by gcp_conn_id.
+    """
+
+    def __init__(self, gcp_conn_id: str = 'google_cloud_default', delegate_to: str = None) -> None:
+        super().__init__(gcp_conn_id, delegate_to=delegate_to)
+        self.num_retries = self._get_field('num_retries', 5)  # type: int
+
+    def get_conn(self):
```

Comment:
nit: there are type annotations in public methods, do we need type hints as well?

Reply:
The Google API Client library (discovery API) does not have valid annotations. Types are automatically generated at runtime.

Reply:
This PR is not intended to make changes to the code. There are no annotations in many places related to GCP, but this month we will try to prepare PR.

Reply:
Got it. Thanks!